File: backend/src/common/error-codes.ts
/// Codes métier STABLES destinés au client Flutter.
/// Le front s'y réfère TOUJOURS par ce code, jamais par le texte du message
/// (voir CONVENTIONS.md § Format de réponse & erreurs).
export enum ErrorCode {
  // Authentification / session
  INVALID_CREDENTIALS = 'INVALID_CREDENTIALS',
  ACCOUNT_DISABLED = 'ACCOUNT_DISABLED',
  PASSWORD_CHANGE_REQUIRED = 'PASSWORD_CHANGE_REQUIRED',
  /// Access token : absent, malformé ou expiré → le client tente un /auth/refresh.
  ACCESS_TOKEN_MISSING = 'ACCESS_TOKEN_MISSING',
  ACCESS_TOKEN_INVALID = 'ACCESS_TOKEN_INVALID',
  REFRESH_TOKEN_INVALID = 'REFRESH_TOKEN_INVALID',
  REFRESH_TOKEN_EXPIRED = 'REFRESH_TOKEN_EXPIRED',
  REFRESH_TOKEN_REVOKED = 'REFRESH_TOKEN_REVOKED',
  FORBIDDEN_ROLE = 'FORBIDDEN_ROLE',
  FORBIDDEN_PERMISSION = 'FORBIDDEN_PERMISSION',
  /// Changement de mot de passe : le mot de passe ACTUEL saisi est faux
  /// (distinct d'INVALID_CREDENTIALS, qui concerne la connexion).
  CURRENT_PASSWORD_INVALID = 'CURRENT_PASSWORD_INVALID',

  // Gestion des comptes
  /// Un admin ne modifie ni ses propres rôles, ni sa propre activation.
  SELF_MODIFICATION_FORBIDDEN = 'SELF_MODIFICATION_FORBIDDEN',
  /// L'opération laisserait le système sans aucun administrateur actif.
  LAST_ACTIVE_ADMIN = 'LAST_ACTIVE_ADMIN',

  // Générique
  VALIDATION_FAILED = 'VALIDATION_FAILED',
  NOT_FOUND = 'NOT_FOUND',
  CONFLICT = 'CONFLICT',
  NOT_IMPLEMENTED = 'NOT_IMPLEMENTED',
  /// Trop de requêtes dans la fenêtre (anti-brute-force) : réessayer plus tard.
  RATE_LIMITED = 'RATE_LIMITED',

  // Stock (règles 2, 9 de CLAUDE.md)
  STOCK_NEGATIVE = 'STOCK_NEGATIVE',
  STOCK_INSUFFICIENT = 'STOCK_INSUFFICIENT',

  // Ventes / facturation (règles 11, 13)
  INVOICE_ONLINE_ONLY = 'INVOICE_ONLINE_ONLY',
  PRICE_NOT_DEFINED = 'PRICE_NOT_DEFINED',
  DISCOUNT_NOT_ALLOWED = 'DISCOUNT_NOT_ALLOWED',
  /// Prix saisi en vente sous le dernier prix d'achat : jamais de vente à perte
  /// (décision MEDMEDBEN 2026-09-22).
  PRICE_BELOW_COST = 'PRICE_BELOW_COST',
  CREDIT_LIMIT_EXCEEDED = 'CREDIT_LIMIT_EXCEEDED',
  /// Total calculé serveur ≠ total annoncé au client (prix changé) : rien écrit.
  SALE_TOTAL_CHANGED = 'SALE_TOTAL_CHANGED',
  /// Id de vente déjà enregistré avec un AUTRE panier : la première vente EXISTE
  /// (stock sorti, espèces en caisse) — ne pas la refaire à l'aveugle.
  SALE_ALREADY_RECORDED = 'SALE_ALREADY_RECORDED',
  /// Id de règlement/paiement déjà enregistré avec un AUTRE contenu : le
  /// premier paiement EXISTE (dette réduite, caisse mouvementée).
  PAYMENT_ALREADY_RECORDED = 'PAYMENT_ALREADY_RECORDED',
  /// Facture demandée sans les mentions légales du magasin (STORE_NIF, STORE_RC).
  STORE_IDENTITY_MISSING = 'STORE_IDENTITY_MISSING',

  // Caisse (règle 12)
  CASH_SESSION_REQUIRED = 'CASH_SESSION_REQUIRED',
  CASH_SESSION_ALREADY_OPEN = 'CASH_SESSION_ALREADY_OPEN',
  /// Sortie de caisse supérieure aux espèces réellement dans le tiroir.
  CASH_INSUFFICIENT = 'CASH_INSUFFICIENT',
  /// Vente hors-ligne antérieure à l'ouverture de la caisse actuelle : ses
  /// espèces appartiennent à une caisse CLÔTURÉE (ou à aucune) — les imputer à
  /// celle-ci fausserait les deux rapports Z.
  CASH_SESSION_CLOSED = 'CASH_SESSION_CLOSED',

  // Catalogue (règle 15)
  BARCODE_ALREADY_USED = 'BARCODE_ALREADY_USED',

  // Inventaire (spec §22)
  /// Un AUTRE inventaire a déjà corrigé ces produits depuis ce comptage : le
  /// même écart serait appliqué deux fois. (Une vente ou une réception, elles,
  /// ne gênent pas : l'ajustement est un delta, il s'y ajoute.)
  INVENTORY_STALE_COUNT = 'INVENTORY_STALE_COUNT',

  // Machines à états
  INVALID_STATE_TRANSITION = 'INVALID_STATE_TRANSITION',

  // Synchronisation offline (docs/context.md)
  SYNC_MUTATION_REJECTED = 'SYNC_MUTATION_REJECTED',
  /// Échec TEMPORAIRE côté serveur : la mutation n'est pas mémorisée, le client
  /// la garde en file et la renverra telle quelle (contrat de sync § idempotence).
  SYNC_RETRY_LATER = 'SYNC_RETRY_LATER',
  /// Le lot déclare un auteur qui n'est pas le porteur de la session : rien
  /// n'est traité ni mémorisé, les mutations restent en file côté appareil.
  SYNC_AUTHOR_MISMATCH = 'SYNC_AUTHOR_MISMATCH',
}

File: app/lib/core/error/error_codes.dart
/// Miroir EXACT de `backend/src/common/error-codes.ts`.
///
/// Le client se base TOUJOURS sur ce code, jamais sur le texte du message
/// (CONVENTIONS.md § Format de réponse & erreurs).
class ErrorCodes {
  const ErrorCodes._();

  // Authentification / session
  static const String invalidCredentials = 'INVALID_CREDENTIALS';
  static const String accountDisabled = 'ACCOUNT_DISABLED';
  static const String passwordChangeRequired = 'PASSWORD_CHANGE_REQUIRED';
  static const String accessTokenMissing = 'ACCESS_TOKEN_MISSING';
  static const String accessTokenInvalid = 'ACCESS_TOKEN_INVALID';
  static const String refreshTokenInvalid = 'REFRESH_TOKEN_INVALID';
  static const String refreshTokenExpired = 'REFRESH_TOKEN_EXPIRED';
  static const String refreshTokenRevoked = 'REFRESH_TOKEN_REVOKED';
  static const String forbiddenRole = 'FORBIDDEN_ROLE';
  static const String forbiddenPermission = 'FORBIDDEN_PERMISSION';
  static const String currentPasswordInvalid = 'CURRENT_PASSWORD_INVALID';

  // Gestion des comptes
  static const String selfModificationForbidden = 'SELF_MODIFICATION_FORBIDDEN';
  static const String lastActiveAdmin = 'LAST_ACTIVE_ADMIN';

  // Générique
  static const String validationFailed = 'VALIDATION_FAILED';
  static const String notFound = 'NOT_FOUND';
  static const String conflict = 'CONFLICT';
  static const String notImplemented = 'NOT_IMPLEMENTED';
  static const String rateLimited = 'RATE_LIMITED';

  // Stock
  static const String stockNegative = 'STOCK_NEGATIVE';
  static const String stockInsufficient = 'STOCK_INSUFFICIENT';

  // Ventes / facturation
  static const String invoiceOnlineOnly = 'INVOICE_ONLINE_ONLY';
  static const String priceNotDefined = 'PRICE_NOT_DEFINED';
  static const String discountNotAllowed = 'DISCOUNT_NOT_ALLOWED';
  static const String creditLimitExceeded = 'CREDIT_LIMIT_EXCEEDED';
  static const String saleTotalChanged = 'SALE_TOTAL_CHANGED';
  static const String saleAlreadyRecorded = 'SALE_ALREADY_RECORDED';
  static const String paymentAlreadyRecorded = 'PAYMENT_ALREADY_RECORDED';

  // Caisse
  static const String cashSessionRequired = 'CASH_SESSION_REQUIRED';
  static const String cashSessionAlreadyOpen = 'CASH_SESSION_ALREADY_OPEN';

  // Catalogue
  static const String barcodeAlreadyUsed = 'BARCODE_ALREADY_USED';

  // Machines à états
  static const String invalidStateTransition = 'INVALID_STATE_TRANSITION';

  // Synchronisation
  static const String syncMutationRejected = 'SYNC_MUTATION_REJECTED';

  /// Échec TEMPORAIRE : la mutation n'est pas mémorisée côté serveur,
  /// le client la GARDE en file et la renverra telle quelle.
  static const String syncRetryLater = 'SYNC_RETRY_LATER';

  /// Codes qui exigent une reconnexion : le refresh ne peut rien y faire.
  static const Set<String> requiresRelogin = {
    refreshTokenInvalid,
    refreshTokenExpired,
    refreshTokenRevoked,
    accountDisabled,
  };

  /// Messages destinés à l'utilisateur, en français, dérivés du code stable.
  static String userMessage(String? code, String fallback) => switch (code) {
    invalidCredentials => 'Identifiant ou mot de passe incorrect',
    currentPasswordInvalid => 'Mot de passe actuel incorrect',
    accountDisabled => 'Ce compte est désactivé',
    passwordChangeRequired => 'Vous devez changer votre mot de passe',
    refreshTokenExpired ||
    refreshTokenRevoked ||
    refreshTokenInvalid => 'Session expirée, reconnectez-vous',
    forbiddenRole ||
    forbiddenPermission => "Vous n'avez pas les droits pour cette action",
    selfModificationForbidden =>
      'Vos rôles et votre activation sont modifiés par un autre administrateur',
    lastActiveAdmin =>
      'Impossible : ce compte est le dernier administrateur actif',
    rateLimited => 'Trop de tentatives. Patientez quelques minutes.',
    stockNegative ||
    stockInsufficient => 'Stock insuffisant pour cette opération',
    cashSessionRequired => 'Ouvrez une session de caisse au préalable',
    invoiceOnlineOnly =>
      'La facturation exige une connexion — le ticket reste valable',
    creditLimitExceeded => 'Limite de crédit du client dépassée',
    notImplemented => "Fonctionnalité pas encore disponible",
    _ => fallback,
  };
}
